### DISCUSSION TITLE
Render bug when visibility changes

### DISCUSSION THREAD

**Discussion by Wupval** [cefsharp/CefSharp#4274] [General]
Hi, we found a bug when using CefSharp for WPF.
When there are multiple browser instances (in our case it starts to bug when there are 6), and a browser goes invisible and visible again, it stops rendering.
For example, when 6 browsers all display the google.com site and the visibility gets toggled off and on, then it is impossible to scroll or click, no animations, etc.
Interestingly, the cursor still changes according to the frozen site and the tooltips also work.

This happens in Nuget CefSharp.Wpf.NETCore version 106.0.260 for x64, but we traced it back to when we upgraded from 93.1.140 to 94.4.20
OS is Windows 10, x64
.NET 6.0, also happens in .NET 5.0

Generally, we noticed it when having only 4 browsers with their Dev-Tools open. After some testing in the Minimal Example, I realised it also happens when there are many browser instances.
I would have reported it as an issue, but I don't actually know if it is a CefSharp bug or a Cef bug and I don't know how to recreate it in C++

If you want to recreate the bug yourself, the files of the example are attached (Just copy paste into the minimal example).
Just start the app, wait for the instances to finish being created, toggle the visibility of the first browser with the button in top right.
[example_files.zip](https://github.com/cefsharp/CefSharp/files/9784214/example_files.zip)

**Comment by amaitland** [comment]
Likely a CEF issue.

https://bitbucket.org/chromiumembedded/cef/issues/2483/osr-invalidate-does-not-generate-frame

**Reply by amaitland** [reply]
Try resizing the window to see if they start redrawing.

**Reply by Wupval** [reply]
Thanks for the fast reply. Unfortunatly, resizing does not work when using the browser in a layout without fixed size (Atleast in our case).
I read the CEF-Issue you linked and the other stated workarounds do not work either, except for calling WasResized() on the BrowserHost after the browser is visible.

Though this would be a workaround, I feel like others would also encounter this problem again. I know that there once was a "hack" for that in the release and it was removed afterwards. Is there a possibility that the workaround gets readded again until the issue gets resolved in CEF?

**Reply by amaitland** [reply]
> Unfortunatly, resizing does not work when using the browser in a layout without fixed size (Atleast in our case).

For testing purposes does resizing the window cause the browsers to resume rendering?

**Reply by Wupval** [reply]
Yes, after resizing the window, the browser renders again

**Reply by amaitland** [reply]
> Is there a possibility that the workaround gets readded again until the issue gets resolved in CEF?

Someone would need to double check the code still works and report back.

It would likely be disabled by default this time.

**Reply by Wupval** [reply]
I could test it, but the earliest I can report back is 2nd of November due to a mix of national holidays and vacation.

**Reply by Wupval** [reply]
I tested the code of the workaround. It seems to work fine in both the minimal example and our application.

**Reply by amaitland** [reply]
Please elaborate on what you actually tested? Did you cherry pick the relevant commit into a specific branch? What did you do exactly? Thanks

**Reply by Wupval** [reply]
My apologies, I forgot to add that. I derived from ChromiumWebBrowser and integrated the changes of the commit https://github.com/cefsharp/CefSharp/commit/9140c08a34e113b8865a80a7ac9f8b4aa19631c4.
As there where code changes since that commit, I did not change anything in OnWindowStateChanged or OnIsVisibleChanged. Instead, I moved the call of ResizeHackFor2779 to OnBrowserWasHidden. I hope this is satisfatory as a test.

**Reply by Wupval** [reply]
Which way would you prefer to have it tested?

**Reply by amaitland** [reply]
As it sounds like it works then next step will be to port the commit to target `master`. PRs welcome.

It's probably also worth creating an new issue on [CEF Issue Tracker](https://bitbucket.org/chromiumembedded/cef/src). The original issue is marked as resolved. Make sure the issue actually gets fixed.

**Reply by Wupval** [reply]
Pull request is created https://github.com/cefsharp/CefSharp/pull/4307
I'll look into creating a new issue for CEF, but this will take some time

**Reply by mayphi** [reply]
Cef Issue: https://github.com/chromiumembedded/cef/issues/3427

**Reply by jingtao01** [reply]
> Likely a CEF issue.
> 
> https://bitbucket.org/chromiumembedded/cef/issues/2483/osr-invalidate-does-not-generate-frame

Also happened here.
And I found that calling browser.GetHost().NotifyScreenInfoChanged() in ChromiumWebBrowser::OnBrowserWasHidden when hidden is false can fix this instead of thre resize hack, because I might change size later.
Just FYI.
